File: chap2/play-jbook/app/comparison/Generics.java
package comparison;

import java.util.List;

class Generics {

    //204 chars
    class NonEmtpyList<G> {
        public G head;
        public List<G> rest;

        public NonEmtpyList(G head, List<G> rest) {
            this.head = head;
            this.rest = rest;
        }
    }


    //351 chars
    interface Ser {
        Byte[] export();
    }
    interface Writeable<El extends Ser> {
        void write(El ser);
    }
    abstract class AdaptableWriter<W extends Writeable<El>, El extends Ser> {
        public W target;

        public abstract El adapt(El e);

        public void out(El e) {
            target.write(adapt(e));
        }
    }


    //simulate a Function... Similar to the Command Pattern
    interface Function1<A,B> {
        B apply(A a);
    }
    //WILL FAIL AT COMPILE TIME
    // interface Functor<F<*>> {
    //     <A, B> F<B> fmap(F<A> a, Function1<A,B> f);
    // }

}

File: chap8/play-sbook/app/comparison/Generics.scala
package comparison



object Generics {

    //48 chars
    case class NonEmptyList[G](h:G, rest:List[G]) {}










    //277 chars
    trait Ser {
        def export:Array[Byte]
    }
    trait Writeable[El <: Ser] {
        def write(el:El)
    }
    trait AdaptableWriter[W <: Writeable[El], El <: Ser] {
        def target:W

        def adapt(el:El):El

        def out(el:El) = target.write(adapt(el))
    }




    //WILL COMPILE -- now, search the web for Functor ;-)
    trait Functor[F[_]] {
        def fmap[A, B](a:F[A], f:A=>B):F[B]
    }

    //an instance of a Functor... Scala provides it for all traversable: map
    object ListFunctor extends Functor[List] {
        def fmap[A,B](a:List[A], f:A=>B) = 
            for (item <- a) yield f(item)
    }

}